### PR TITLE
[csi] add snapshot support

### PR DESCRIPTION
**Pull request by jichenjc**
Add snapshot function for csi including:
add snapshot
delete snapshot
list snapshot

**What this PR does / why we need it**:
add snapshot functions for csi cinder volume
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
not a bug fix
**Special notes for your reviewer**:

**Release note**:

```
add snapshot csi functions, create-snapshot, delete-snapshot and list-snapshot are added 
```

### CODE REVIEW

File: pkg/csi/cinder/controllerserver.go [hunk 1]
```diff
@@ -221,5 +221,88 @@ func (cs *controllerServer) ListVolumes(ctx context.Context, req *csi.ListVolume
...
+
+	name := req.Name
+	volumeId := req.SourceVolumeId
+	// No description from csi.CreateSnapshotRequest now
+	description := ""
+
+	// Delegate the check to openstack itself
+	snap, err := cloud.CreateSnapshot(name, volumeId, description, &req.Parameters)
+	if err != nil {
+		glog.V(3).Infof("Faled to Create snapshot: %v", err)
```
Comment (adisky) [inline pkg/csi/cinder/controllerserver.go]: s/faled/failed

File: pkg/csi/cinder/controllerserver.go [hunk 2]
```diff
@@ -221,5 +221,88 @@ func (cs *controllerServer) ListVolumes(ctx context.Context, req *csi.ListVolume
...
+		return nil, err
+	}
+
+	name := req.Name
+	volumeId := req.SourceVolumeId
+	// No description from csi.CreateSnapshotRequest now
+	description := ""
+
+	// Delegate the check to openstack itself
+	snap, err := cloud.CreateSnapshot(name, volumeId, description, &req.Parameters)
```
Comment (adisky) [inline pkg/csi/cinder/controllerserver.go]: 1. various checks missing here, like if CO request to create a snapshot that already exists, It should not respond any error and return the details of existing snapshot.
2. Snapshot with same name on two different volume-id not allowed.

If you prefer to implement this checks later, add in todo as
TODO: Idempotency Checks

Comment (adisky) [reply pkg/csi/cinder/controllerserver.go]: @jichenjc for q2 i know it is allowed on openstack side, but after reading csi specs, i think it is not allowed, for q1 you should not respond with any error [It is clearly mentioned in csi specs]

Comment (jichenjc) [reply pkg/csi/cinder/controllerserver.go]: ok, I will add more to do and some checks here..

**Comment by adisky** [comment]
@jichenjc can we add snapshot tests in CI also??

**Comment by dims** [comment]
@adisky @flaper87 this looks ready?

**Pull request merged**